### ISSUE TITLE
Generator + Scaffold Support for MongoDB

### ISSUE THREAD

**Issue by thedavidprice** [redwoodjs/graphql#3077]
Prisma now offers MongoDB support as a preview feature. See [release notes](https://github.com/prisma/prisma/releases/tag/2.27.0)

Based on my initial test using MongoDB with the Redwood Tutorial, the changes were trivial:
- I used the scaffold generators
- then updated the model `id` fields in both the schema.prisma and SDLs
    - code here: https://github.com/thedavidprice/redwood-tutorial-mongo/blob/main/api/db/schema.prisma

Next step is to determine what it will take to support MongoDB and then prioritize against Redwood v1.

**Comment by dthyresson** [comment]
Is there a way to detect the data provider (SQLite, Postgres, MongoDB) such that the generators can be smart and adjust the output accordingly -- or would the generator need to specify "mongodb" via a flag.

The former, to detect, would be nicer DX.

**Comment by thedavidprice** [comment]
Yes, we used the (undocumented) api helper in the Render setup command.

**Comment by thedavidprice** [comment]
## Help Wanted
If someone has time, it would be great to know the current state of Redwood Generators supporting MongoDB as is. We might actually be in a very good place already... TBD

**Comment by ajcwebdev** [comment]
Scaffold seemed to work flawlessly.

![01-redwood-admin-with-mongodb](https://dev-to-uploads.s3.amazonaws.com/uploads/articles/yxfxm54a7n17dezag4pw.png)

![02-mongodb-atlas-with-seed-data](https://dev-to-uploads.s3.amazonaws.com/uploads/articles/gf772mp6umqgrb71f5r2.png)

## Resources

* Check the example repo [here](https://github.com/ajcwebdev/redwood-mongo).
* If you need help getting setup with Atlas you can view my guide [here](https://dev.to/ajcwebdev/can-i-use-mongodb-with-prisma-yet-50go).

## Prisma Schema

```prisma
datasource db {
  provider = "mongodb"
  url      = env("DATABASE_URL")
}

generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["mongodb"]
}

model Post {
  id    String @id @default(dbgenerated()) @map("_id") @db.ObjectId
  title String
  body  String
}
```

**Comment by dthyresson** [comment]
@ajcwebdev Did the Redwood generators (sdl, service, or scaffold) create this line:

```js
 id    String @id @default(dbgenerated()) @map("_id") @db.ObjectId
```
?

**Comment by ajcwebdev** [comment]
Negative, that was from the original Prisma documentation back in April. Feel free to try out other ways of generating the `id`.

**Comment by dthyresson** [comment]
> Negative, that was from the original Prisma documentation back in April. Feel free to try out other ways of generating the `id`.

Right. :) Yeah, I deleted my comment right after I posted when I realized I was looking at the model and not sdl.

**Comment by ajcwebdev** [comment]
Next step for me will be trying this out with [MongoDB on Railway](https://docs.railway.app/plugins/mongodb) and writing up a forum post, Atlas is a huge drag.

**Comment by dthyresson** [comment]
> writing up a forum post,

Docs like these too instead of Postgres would help: https://redwoodjs.com/docs/local-postgres-setup

Also, curious about Prisma's shadow db and testing - does Redwood's unit tests and mocks/scenarios work well.

But I assume if https://redwoodjs.com/docs/testing#the-test-database is also Mongo, then should be ok.

**Comment by ajcwebdev** [comment]
Cool, can definitely write some docs for local setup and will try out the tests as well.

**Comment by nikolasburk** [comment]
MongoDB support in Prisma is now production-ready 🎉 https://www.prisma.io/blog/mongodb-general-availability-pixnun6mffmu
